File: apps/api/routers/knowledge.py
"""
知识图谱路由
P4-05: 图谱API
"""

from typing import List, Dict

from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from packages.db import Tenant, User
from services.knowledge import KnowledgeGraphService, LearningService

from ..deps import get_db, get_current_tenant, get_current_user

router = APIRouter()


@router.get("/graph")
async def get_knowledge_graph(
    tenant: Tenant = Depends(get_current_tenant),
    db: Session = Depends(get_db),
):
    """获取知识图谱"""
    service = KnowledgeGraphService(db)
    graph = service.build_graph(tenant.id)
    return graph.to_dict()


@router.get("/concepts/{concept}/videos")
async def get_concept_videos(
    concept: str,
    tenant: Tenant = Depends(get_current_tenant),
    db: Session = Depends(get_db),
):
    """获取包含某概念的视频列表"""
    service = KnowledgeGraphService(db)
    videos = service.get_concept_videos(tenant.id, concept)
    return [
        {
            "id": v.id,
            "bvid": v.bvid,
            "title": v.title,
            "author": v.author,
        }
        for v in videos
    ]


@router.get("/concepts/{concept}/related")
async def get_related_concepts(
    concept: str,
    limit: int = 10,
    tenant: Tenant = Depends(get_current_tenant),
    db: Session = Depends(get_db),
):
    """获取相关概念"""
    service = KnowledgeGraphService(db)
    return service.get_related_concepts(tenant.id, concept, limit)


@router.get("/learning/stats")
async def get_learning_stats(
    days: int = 7,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """获取学习统计"""
    service = LearningService(db)
    stats = service.get_user_stats(user.id, days)
    
    return {
        "total_videos": stats.total_videos,
        "total_duration": stats.total_duration,
        "completed_videos": stats.completed_videos,
        "concepts_learned": stats.concepts_learned,
        "videos_by_day": stats.videos_by_day,
        "top_authors": stats.top_authors,
    }


@router.get("/learning/weekly-report")
async def get_weekly_report(
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """获取周报"""
    service = LearningService(db)
    report = service.generate_weekly_report(user.id)
    
    return {
        "week_start": report.week_start.isoformat(),
        "week_end": report.week_end.isoformat(),
        "highlights": report.highlights,
        "stats": {
            "total_videos": report.stats.total_videos,
            "total_duration": report.stats.total_duration,
            "concepts_learned": report.stats.concepts_learned,
        },
        "recommendations": report.recommendations,
    }


@router.get("/learning/review-suggestions")
async def get_review_suggestions(
    limit: int = 5,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """获取复习建议"""
    service = LearningService(db)
    return service.get_review_suggestions(user.id, limit)

File: apps/web/src/lib/api/knowledge.ts
/**
 * 知识图谱 API
 */

import { client } from "./client"

// 类型定义
export interface GraphNode {
  id: string
  label: string
  type: "concept" | "video"
  size?: number
  metadata?: Record<string, unknown>
}

export interface GraphEdge {
  source: string
  target: string
  weight: number
}

export interface KnowledgeGraph {
  nodes: GraphNode[]
  edges: GraphEdge[]
  stats: {
    total_concepts: number
    total_videos: number
    total_edges: number
  }
}

export interface ConceptVideo {
  id: number
  source_type: string
  source_id: string
  bvid?: string  // 兼容字段
  title: string
  author: string
}

export interface RelatedConcept {
  concept: string
  weight: number
  videos_count: number
}

export interface LearningStats {
  total_videos: number
  total_duration: number
  completed_videos: number
  concepts_learned: number
  videos_by_day: Record<string, number>
  top_authors: Array<{ author: string; count: number }>
}

export interface WeeklyReport {
  week_start: string
  week_end: string
  highlights: string[]
  stats: {
    total_videos: number
    total_duration: number
    concepts_learned: number
  }
  recommendations: string[]
}

export interface ReviewSuggestion {
  video_id: number
  source_type: string
  source_id: string
  bvid?: string  // 兼容字段
  title: string
  concepts: string[]
  last_viewed: string
  priority: number
}

// API 方法
export const knowledgeApi = {
  /**
   * 获取知识图谱
   */
  async getGraph() {
    return client.get<KnowledgeGraph>("/knowledge/graph")
  },

  /**
   * 获取概念关联的视频
   */
  async getConceptVideos(concept: string) {
    return client.get<ConceptVideo[]>(`/knowledge/concepts/${encodeURIComponent(concept)}/videos`)
  },

  /**
   * 获取相关概念
   */
  async getRelatedConcepts(concept: string, limit = 10) {
    return client.get<RelatedConcept[]>(
      `/knowledge/concepts/${encodeURIComponent(concept)}/related`,
      { params: { limit } }
    )
  },

  /**
   * 获取学习统计
   */
  async getLearningStats(days = 7) {
    return client.get<LearningStats>("/knowledge/learning/stats", { params: { days } })
  },

  /**
   * 获取周报
   */
  async getWeeklyReport() {
    return client.get<WeeklyReport>("/knowledge/learning/weekly-report")
  },

  /**
   * 获取复习建议
   */
  async getReviewSuggestions(limit = 5) {
    return client.get<ReviewSuggestion[]>("/knowledge/learning/review-suggestions", { params: { limit } })
  },
}
